overlord/state/change.go

```diff
@@ -282,32 +296,128 @@ func init() {
 	}
 }
 
+func (c *Change) isTaskWaiting(visited map[string]taskWaitComputeStatus, t *Task, deps []*Task) bool {
+	taskID := t.ID()
+	// Retrieve the compute status of the wait for the task, if not
+	// computed this defaults to 0 (taskWaitStatusNotComputed).
+	computeStatus := visited[taskID]
+	switch computeStatus {
+	case taskWaitStatusComputing:
+		// Cyclic dependency detected, return false to short-circuit.
+		logger.Noticef("detected cyclic dependencies for task %q in change %q", t.Kind(), t.Change().Kind())
+		return false
+	case taskWaitStatusWaiting, taskWaitStatusNotWaiting:
+		return computeStatus == taskWaitStatusWaiting
+	}
+	visited[taskID] = taskWaitStatusComputing
+
+	var isWaiting bool
+depscheck:
+	for _, wt := range deps {
+		switch wt.Status() {
+		case WaitStatus:
+			isWaiting = true
+		// States that can are valid when waiting
+		// - Done, Undone
+		case DoneStatus, UndoneStatus:
+			continue
+		// For 'Do' and 'Undo' we have to check whether the task is waiting
+		// for any dependencies. The logic is the same, but the set of tasks
+		// varies.
+		case DoStatus:
+			isWaiting = c.isTaskWaiting(visited, wt, wt.WaitTasks())
+			if !isWaiting {
+				// Cancel early if we detect something is runnable.
+				break depscheck
+			}
+		case UndoStatus:
+			isWaiting = c.isTaskWaiting(visited, wt, wt.HaltTasks())
+			if !isWaiting {
+				// Cancel early if we detect something is runnable.
+				break depscheck
+			}
+		default:
+			// When we determine the change can not be in a wait-state then
+			// break early.
+			isWaiting = false
+			break depscheck
+		}
+	}
+	if isWaiting {
+		visited[taskID] = taskWaitStatusWaiting
+	} else {
+		visited[taskID] = taskWaitStatusNotWaiting
+	}
+	return isWaiting
+}
+
+// isChangeWaiting should only ever return true iff it determines all tasks in Do/Undo
+// are blocked by tasks in either of three states: 'DoneStatus', 'UndoneStatus' or 'WaitStatus',
+// if this fails, we default to the normal status ordering logic.
+func (c *Change) isChangeWaiting() bool {
+	// Since we might visit tasks more than once, we store results to avoid recomputing them.
+	visited := make(map[string]taskWaitComputeStatus)
+	for _, t := range c.Tasks() {
+		switch t.Status() {
+		case WaitStatus, DoneStatus, UndoneStatus:
```

Comment:
Following @pedronis 's comment this line should include also Error and Hold statuses